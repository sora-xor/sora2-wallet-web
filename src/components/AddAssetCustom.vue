<template>
  <div class="asset-custom">
    <s-input
      ref="search"
      class="asset-custom-address"
      :maxlength="128"
      :placeholder="t(`addAsset.${AddAssetTabs.Custom}.addressPlaceholder`)"
      border-radius="mini"
      v-model="address"
      @change="handleSearch"
    />
    <div v-if="address && !selectedAsset" class="asset-custom-empty">
      {{ t(`addAsset.${alreadyAttached ? 'alreadyAttached' : 'empty'}`) }}
    </div>
    <s-button type="primary" :disabled="!(selectedAsset && address)" @click="navigateToAddAssetDetails">
      {{ t('addAsset.next') }}
    </s-button>
  </div>
</template>

<script lang="ts">
import { Component, Mixins } from 'vue-property-decorator'
import { Getter, Action } from 'vuex-class'
import { AccountAsset, Asset } from '@sora-substrate/util'

import TranslationMixin from './mixins/TranslationMixin'
import { AddAssetTabs, RouteNames } from '../consts'

@Component
export default class AddAssetCustom extends Mixins(TranslationMixin) {
  readonly AddAssetTabs = AddAssetTabs

  @Getter accountAssets!: Array<AccountAsset>
  @Action navigate!: (options: { name: string; params?: object }) => Promise<void>
  @Action searchAsset!: (options: { address: string }) => Promise<Asset>

  address = ''
  selectedAsset: Asset | null = null
  alreadyAttached = false

  mounted (): void {
    const input = this.$refs.search as any

    if (input && typeof input.focus === 'function') {
      input.focus()
    }
  }

  async handleSearch (value: string): Promise<void> {
    this.alreadyAttached = false
    if (!value.trim()) {
      this.selectedAsset = null
      return
    }
    const search = value.trim().toLowerCase()
    const asset = await this.searchAsset({ address: search })
    if (this.accountAssets.find(({ address }) => address.toLowerCase() === search)) {
      this.selectedAsset = null
      this.alreadyAttached = true
      return
    }
    this.selectedAsset = asset
    if (!this.selectedAsset?.symbol) {
      this.selectedAsset = null
    }
  }

  navigateToAddAssetDetails (): void {
    this.navigate({ name: RouteNames.AddAssetDetails, params: { asset: this.selectedAsset } })
  }
}
</script>

<style scoped lang="scss">
.asset-custom {
  > * {
    margin-top: $basic-spacing;
  }
  &-empty {
    @include hint-text;
  }
  .el-button--primary {
    width: 100%;
  }
}
</style>
