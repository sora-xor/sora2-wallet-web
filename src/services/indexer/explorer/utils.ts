import { formatStringNumber } from '../../../util';

import type { SubqueryAssetEntity } from '../subquery/types';
import type { SubsquidAssetEntity } from '../subsquid/types';
import type { FiatPriceObject, UpdatesStream, PoolApyObject } from '../types';

export function parseAssetFiatPrice(entity: SubsquidAssetEntity | SubqueryAssetEntity): FiatPriceObject {
  const acc = {};
  const id = entity.id;
  const priceFPNumber = formatStringNumber(entity.priceUSD);
  const isPriceFinity = priceFPNumber.isFinity();
  if (isPriceFinity) {
    acc[id] = priceFPNumber.toCodecString();
  }
  return acc;
}

export function parsePriceStreamUpdate(entity: UpdatesStream): Nullable<FiatPriceObject> {
  if (!entity?.data) return null;

  const data = JSON.parse(entity.data);

  return Object.entries(data).reduce((acc, [id, price]) => {
    const priceFPNumber = formatStringNumber(price as string);
    const isPriceFinity = priceFPNumber.isFinity();
    if (isPriceFinity) {
      acc[id] = priceFPNumber.toCodecString();
    }
    return acc;
  }, {});
}

export function parseApyStreamUpdate(entity: UpdatesStream): PoolApyObject {
  const data = entity?.data ? JSON.parse(entity.data) : {};

  return Object.entries(data).reduce((acc, [id, apy]) => {
    const strategicBonusApyFPNumber = formatStringNumber(apy as string);
    const isStrategicBonusApyFinity = strategicBonusApyFPNumber.isFinity();
    if (isStrategicBonusApyFinity) {
      acc[id] = strategicBonusApyFPNumber.toCodecString();
    }
    return acc;
  }, {});
}
