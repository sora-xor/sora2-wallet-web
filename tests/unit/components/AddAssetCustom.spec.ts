import Vuex from 'vuex'
import { shallowMount, createLocalVue } from '@vue/test-utils'

import AddAssetCustom from '@/components/AddAssetCustom.vue'
import { TranslationMock, SoramitsuElementsImport } from '../../utils'

const localVue = createLocalVue()
localVue.use(Vuex)
const store = new Vuex.Store({
  modules: {
    Router: {
      actions: {
        navigate: jest.fn()
      }
    },
    Account: {
      actions: {
        searchAsset: jest.fn(),
        addAsset: jest.fn()
      }
    }
  } as any
})

describe('AddAssetCustom.vue', () => {
  beforeEach(() => {
    SoramitsuElementsImport(localVue)
    TranslationMock(AddAssetCustom)
  })

  it('should renders correctly', () => {
    const wrapper = shallowMount(AddAssetCustom, { localVue, store })
    expect(wrapper.element).toMatchSnapshot()
  })
})
