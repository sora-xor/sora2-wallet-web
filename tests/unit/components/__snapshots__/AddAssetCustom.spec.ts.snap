// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddAssetCustom.vue should renders correctly 1`] = `
<div
  class="asset-custom"
>
  <s-input-stub
    accept=""
    autocomplete="off"
    borderradius="mini"
    class="asset-custom-address"
    form=""
    label=""
    maxlength="128"
    name=""
    placeholder=""
    prefix=""
    size="big"
    suffix=""
    tabindex=""
    type="text"
    value=""
  />
   
  <!---->
   
  <!---->
   
  <div
    class="asset-custom-actions s-flex"
  >
    <s-button-stub
      borderradius="small"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="secondary"
    >
      
    </s-button-stub>
     
    <s-button-stub
      borderradius="small"
      disabled="true"
      icon=""
      iconposition="left"
      nativetype="button"
      size="medium"
      tabindex="0"
      tooltip=""
      tooltipplacement="top"
      type="primary"
    >
      
      
    
    </s-button-stub>
  </div>
</div>
`;
